com.oracle.truffle.r.ffi.impl/src/com/oracle/truffle/r/ffi/impl/nodes/IsSortedNode.java

```diff
@@ -0,0 +1,119 @@
+ * 3 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.r.ffi.impl.nodes;
+
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.ImportStatic;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.r.runtime.DSLConfig;
+import com.oracle.truffle.r.runtime.data.RDoubleVector;
+import com.oracle.truffle.r.runtime.data.RIntVector;
+import com.oracle.truffle.r.runtime.data.RLogicalVector;
+import com.oracle.truffle.r.runtime.data.RStringVector;
+import com.oracle.truffle.r.runtime.data.VectorDataLibrary;
+import com.oracle.truffle.r.runtime.data.altrep.AltrepSortedness;
+import com.oracle.truffle.r.runtime.data.model.RAbstractAtomicVector;
+
+/**
+ * Represents all possible *_IS_SORTED upcalls eg. INTEGER_IS_SORTED. Note that all these upcalls
+ * just dispatches to {@code VectorDataLibrary.isSorted} message, therefore, we can merge them in
+ * this class.
+ */
+@GenerateUncached
+@ImportStatic(DSLConfig.class)
+public abstract class IsSortedNode extends FFIUpCallNode.Arg1 {
+
+    public static IsSortedNode create() {
+        return IsSortedNodeGen.create();
+    }
+
+    @Specialization(limit = "getTypedVectorDataLibraryCacheSize()")
+    public int isIntSorted(RIntVector intVec,
+                    @CachedLibrary("intVec.getData()") VectorDataLibrary dataLibrary) {
+        return isVectorSorted(intVec, dataLibrary);
+    }
+
+    @Specialization(replaces = "isIntSorted")
+    public int isIntSortedUncached(RIntVector intVec) {
+        return isVectorSorted(intVec, VectorDataLibrary.getFactory().getUncached(intVec.getData()));
+    }
+
+    @Specialization(limit = "getTypedVectorDataLibraryCacheSize()")
+    public int isDoubleSorted(RDoubleVector doubleVector,
+                    @CachedLibrary("doubleVector.getData()") VectorDataLibrary dataLibrary) {
+        return isVectorSorted(doubleVector, dataLibrary);
+    }
+
+    @Specialization(replaces = "isDoubleSorted")
+    public int isDoubleSortedUncached(RDoubleVector doubleVector) {
+        return isVectorSorted(doubleVector, VectorDataLibrary.getFactory().getUncached(doubleVector.getData()));
+    }
+
+    @Specialization(limit = "getTypedVectorDataLibraryCacheSize()")
+    public int isLogicalSorted(RLogicalVector logicalVector,
+                    @CachedLibrary("logicalVector.getData()") VectorDataLibrary dataLibrary) {
+        return isVectorSorted(logicalVector, dataLibrary);
+    }
+
+    @Specialization(replaces = "isLogicalSorted")
+    public int isLogicalSortedUncached(RLogicalVector logicalVector) {
```

Comment:
This is not necessary (and it's probably dead code anyway). This is exactly what the @CachedLibrary already does for you: after the cache is filled, it falls back to the uncached library.